client_plugin/drivers/shared/shared_driver.go

```diff
@@ -402,8 +401,171 @@ func (d *VolumeDriver) Path(r volume.Request) volume.Response {
 func (d *VolumeDriver) Mount(r volume.MountRequest) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Mounting volume ")
 
-	log.Errorf("VolumeDriver Mount to be implemented")
-	return volume.Response{Err: ""}
+	// lock the state
+	d.RefCounts.StateMtx.Lock()
+	defer d.RefCounts.StateMtx.Unlock()
+
+	// checked by refcounting thread until refmap initialized
+	// useless after that
+	d.RefCounts.MarkDirty()
+
+	return d.processMount(r)
+}
+
+// processMount -  process a mount request
+func (d *VolumeDriver) processMount(r volume.MountRequest) volume.Response {
+	d.MountIDtoName[r.ID] = r.Name
+
+	// If the volume is already mounted , just increase the refcount.
+	// Note: for new keys, GO maps return zero value, so no need for if_exists.
+	refcnt := d.IncrRefCount(r.Name) // save map traversal
+	log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
+	if refcnt > 1 {
+		log.WithFields(
+			log.Fields{"name": r.Name, "refcount": refcnt},
+		).Info("Already mounted, skipping mount. ")
+		return volume.Response{Mountpoint: d.GetMountPoint(r.Name)}
+	}
+
+	if plugin_utils.AlreadyMounted(r.Name, d.MountRoot) {
+		log.WithFields(log.Fields{"name": r.Name}).Info("Already mounted, skipping mount. ")
+		return volume.Response{Mountpoint: d.GetMountPoint(r.Name)}
+	}
+
+	mountpoint, err := d.MountVolume(r.Name, "", "", false, true)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"name": r.Name,
+				"error": err},
+		).Error("Failed to mount ")
+
+		refcnt, _ := d.DecrRefCount(r.Name)
+		if refcnt == 0 {
+			log.Infof("Detaching %s - it is not used anymore", r.Name)
+			// TODO: umount here
+		}
+		return volume.Response{Err: err.Error()}
+	}
+
+	return volume.Response{Mountpoint: mountpoint}
+}
+
+// MountVolume - Request attach and then mounts the volume.
+func (d *VolumeDriver) MountVolume(name string, fstype string, id string, isReadOnly bool, skipAttach bool) (string, error) {
+	mountpoint := d.GetMountPoint(name)
+	// First, make sure  that mountpoint exists.
+	err := fs.Mkdir(mountpoint)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"name": name,
+				"dir": mountpoint},
+		).Error("Failed to make directory for volume mount ")
+		return mountpoint, err
+	}
+
+	// Increase GRef
+	log.Infof("Before AtomicIncr")
+	err = d.kvStore.AtomicIncr(kvstore.VolPrefixGRef + name)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"name": name,
+				"error": err},
+		).Error("Failed to increase global refcount when processMount ")
+		return "", err
+	}
+
+	// Blocking wait till Mounted
+	info, err := d.kvStore.BlockingWaitAndGet(kvstore.VolPrefixState+name,
+		string(kvstore.VolStateMounted), kvstore.VolPrefixInfo+name)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to blocking wait for Mounted state. Error: %v.", err)
```

Comment:
No it failed to wait for the change of state.
This caller doesn't do the transit. It's "waiting" for the event handler on the manager nodes to do the transit.
It could fail because timeout (event handler fails to update the state somehow) or because of error (etcd is not available).